locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsUtil.java

```diff
@@ -354,30 +355,60 @@ public static String getFirstReachableAddress(ComputeServiceContext context, Nod
     }
 
+    /** @see #getReachableAddresses(NodeMetadata, Duration, Predicate) */
     public static String getFirstReachableAddress(NodeMetadata node, Duration timeout, Predicate<? super HostAndPort> socketTester) {
+        Iterable<HostAndPort> addresses = getReachableAddresses(node, timeout, socketTester);
+        HostAndPort address = Iterables.getFirst(addresses, null);
+        if (address != null) {
+            return address.getHostText();
+        } else {
+            throw new IllegalStateException("No reachable IPs for " + node + "; check whether the node is " +
+                    "reachable and whether it meets the requirements of the HostAndPort tester: " + socketTester);
+        }
+    }
+
+    /**
+     * @return The public and private addresses of node that are reachable.
+     * @see #getReachableAddresses(Iterable, Duration, Predicate)
+     */
+    public static Iterable<HostAndPort> getReachableAddresses(NodeMetadata node, Duration timeout, Predicate<? super HostAndPort> socketTester) {
         final int port = node.getLoginPort();
-        List<HostAndPort> sockets = FluentIterable
-                .from(Iterables.concat(node.getPublicAddresses(), node.getPrivateAddresses()))
+        return getReachableAddresses(Iterables.concat(node.getPublicAddresses(), node.getPrivateAddresses()), port, timeout, socketTester);
+    }
+
+    /** @see #getReachableAddresses(Iterable, Duration, Predicate) */
+    public static Iterable<HostAndPort> getReachableAddresses(Iterable<String> hosts, final int port, Duration timeout, Predicate<? super HostAndPort> socketTester) {
+        FluentIterable<HostAndPort> sockets = FluentIterable
+                .from(hosts)
                 .transform(new Function<String, HostAndPort>() {
                         @Override public HostAndPort apply(String input) {
                             return HostAndPort.fromParts(input, port);
-                        }})
-                .toList();
-        
+                        }});
+        return getReachableAddresses(sockets, timeout, socketTester);
+    }
+
+    /**
+     * Uses {@link ReachableSocketFinder} to determine which sockets are reachable. Iterators
+     * are unmodifiable and are lazily evaluated.
+     * @param sockets The host-and-ports to test
+     * @param timeout Max time to try to connect to the ip:port
+     * @param socketTester A predicate determining reachability.
+     */
+    public static Iterable<HostAndPort> getReachableAddresses(Iterable<HostAndPort> sockets, Duration timeout, Predicate<? super HostAndPort> socketTester) {
         ListeningExecutorService executor = MoreExecutors.listeningDecorator(Executors.newCachedThreadPool());
         try {
             ReachableSocketFinder finder = new ReachableSocketFinder(socketTester, executor);
-            HostAndPort result = finder.findOpenSocketOnNode(sockets, timeout);
-            return result.getHostText();
-        } catch (Exception e) {
```

Comment:
The `catch` here is useful for adding context to the exception - the node. Could be useful when troubleshooting and cross-referncing logs.